### PR TITLE
Issue-196 Create Jenkins configuration using OpenStack Jenkins Job Builder

### PR DESCRIPTION
**Pull request by nicoloboschi**
Descriptions of the changes in this PR:

Introduction of Jenkins Job Builder script, in order to have the configuration of Jenkins jobs directly stored in GIT repo.
This will enable reviews of Jenkins configuration

### CODE REVIEW

File: jenkins/bookkeeper-master-job-configuration.yaml [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+        - "*/master"
+- job:
+    name: bookkeeper-master
+    description: "Nightly build for BookKeeper master.\r\nThis job is built using jenkins-jobs build scripts committed on BookKeeper repo.\r\n\r\nNo not edit manually this job"
+    project-type: maven
+    jdk: "JDK 1.8 (latest)"
+    node: Hadoop
+    maven:
+      root-pom: pom.xml
+      goals: clean apache-rat:check package findbugs:check -Dmaven.test.failure.ignore=false deploy
```
Comment (sijie) [inline jenkins/bookkeeper-master-job-configuration.yaml]: I am wondering why do we need 'deploy' here. Is it the job publishing SNAPSHOT artifact?

Comment (eolivelli) [inline jenkins/bookkeeper-master-job-configuration.yaml]: This is the.current configuration.
Maybe we should create two jobs ?
 I think @nicoloboschi will add the.job for pr

File: jenkins/bookkeeper-master-job-configuration.yaml [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+    wrappers:
+    - timeout:
+        type: likely-stuck
+        write-description: true
+    scm:
+    - repo
+    properties:
+    - build-discarder:
+        days-to-keep: 14
+        num-to-keep: 5
```
Comment (jiazhai) [inline jenkins/bookkeeper-master-job-configuration.yaml]: Is it allowed to set a bigger num-to-keep number? That would be helpful when meet problems.

File: jenkins/bookkeeper-master-job-configuration.yaml (hunk 1 above)
Comment (nicoloboschi) [inline jenkins/bookkeeper-master-job-configuration.yaml]: @sijie If you want, I can create three job configurations:
- Clean install
- Deploy
- PR builder

File: jenkins/bookkeeper-master-job-configuration.yaml (hunk 2 above)
Comment (nicoloboschi) [inline jenkins/bookkeeper-master-job-configuration.yaml]: @jiazhai I think this issue represents the first step in order to keep stored job configuration, maybe is better to start with a configuration and then improve it

File: jenkins/bookkeeper-master-job-configuration.yaml [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+---
+- scm:
+    name: repo
+    scm:
+    - git:
+        url: https://git-wip-us.apache.org/repos/asf/bookkeeper.git
```
Comment (eolivelli) [inline jenkins/bookkeeper-master-job-configuration.yaml]: @nicoloboschi we have just moved to full github workflow, do you mind changing this to:
https://github.com/apache/bookkeeper.git

**Comment by eolivelli** [comment]
@sijie I have tested the script on my companies Jenkins and it is working as expected.
Do you want to try out locally ?

Actual configuration of the master job is
https://builds.apache.org/job/bookkeeper-master/config.xml

The proposal from @nicoloboschi is create a "maven project job" but actually we have a "freestyle project"

**Comment by sijie** [comment]
@eolivelli @nicoloboschi can you document and provide instructions on how to verify this?

**Comment by eolivelli** [comment]
@sijie I think that the readme file would be enough to start.
It is important to use actual version, the trunk, of jenkins job builder, the latest released version is somehow buggy and does handle the email reporting plugin well

File: jenkins/README.md [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+* Jenkins Job Builder (see https://docs.openstack.org/infra/jenkins-job-builder/)
+
+
+The jenkins folder contains:
+
+ - bookkeeper-master-job-configuration.yaml   (Jenkins job definition)
+ - jenkins_jobs.ini   (Jenkins job builder configuration)
+
+
+Build and install latest version of jenkins-job-builder;
```
Comment (sijie) [inline jenkins/README.md]: Can we skip the build instructions? pip provides installation, right?

File: jenkins/README.md [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+```
+
+How do I create job ?
+
+Get an username and an "API Token" for your Jenkins configuration
+
+In order to test jenkins job configuration, run:
+
+```
+   $ jenkins-jobs test bookkeeper-master-job-configuration.yaml
```
Comment (sijie) [inline jenkins/README.md]: please document which directory to run this command.

File: jenkins/bookkeeper-master-job-configuration.yaml (hunk 3 above)
Comment (sijie) [inline jenkins/bookkeeper-master-job-configuration.yaml]: @nicoloboschi can you change this to the new repo address?

File: jenkins/bookkeeper-master-job-configuration.yaml (hunk 1 above)
Comment (sijie) [inline jenkins/bookkeeper-master-job-configuration.yaml]: I am fine with current master job for now. we can change it later.

File: jenkins/README.md (hunk 4 above)
Comment (nicoloboschi) [inline jenkins/README.md]: Pip downloads 1.6.2 jenkins jobs builder version, which does not support email notification after build. The right version to use is 2.0.0

Comment (sijie) [inline jenkins/README.md]: gotcha. thx

**Comment by sijie** [comment]
@eolivelli +1 thank you

**Pull request closed without merging**